### PR TITLE
[SPARK-4361][Doc] Add more docs for Hadoop Configuration

### PR DESCRIPTION
**Pull request by zsxwing**
I'm trying to point out reusing a Configuration in these APIs is dangerous. Any better idea?

### CODE REVIEW

**Comment by srowen** [comment]
Is the outcome here that the doc changes are OK to merge? documenting current behavior seems good.

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 1]
```diff
@@ -242,7 +242,11 @@ class SparkContext(config: SparkConf) extends SparkStatusAPI with Logging {
   // the bound port to the cluster manager properly
   ui.foreach(_.bind())
 
-  /** A default Hadoop Configuration for the Hadoop code (e.g. file systems) that we reuse. */
+  /** A default Hadoop Configuration for the Hadoop code (e.g. file systems) that we reuse.
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: really small nit but this should be scaladocs instead of javadocs

**Comment by andrewor14** [comment]
I think it's safe to say that we won't implement the alternative behavior that @JoshRosen suggested by the release. For this reason I think we should at least document these unexpected behavior for 1.3 in addition to delaying the fix till later. I'm going to merge this into master and 1.3.

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 2]
```diff
@@ -630,7 +634,10 @@ class SparkContext(config: SparkConf) extends SparkStatusAPI with Logging {
    * necessary info (e.g. file name for a filesystem-based dataset, table name for HyperTable),
    * using the older MapReduce API (`org.apache.hadoop.mapred`).
    *
-   * @param conf JobConf for setting up the dataset
+   * @param conf JobConf for setting up the dataset. Note: This will be put into a Broadcast.
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: Nice find. It seems perfectly reasonable from the user's perspective to just save `sc.hadoopConfiguration` into a val and use it for many things. That's probably what I would have done if I didn't know about the nuances here.

**Pull request closed without merging**